File: java/src/com/v2soft/styxlib/library/server/vfs/MemoryStyxDirectory.java
package com.v2soft.styxlib.library.server.vfs;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import org.apache.mina.core.buffer.IoBuffer;

import com.v2soft.styxlib.library.exceptions.StyxErrorMessageException;
import com.v2soft.styxlib.library.io.StyxByteBufferWriteable;
import com.v2soft.styxlib.library.messages.base.enums.FileMode;
import com.v2soft.styxlib.library.messages.base.enums.ModeType;
import com.v2soft.styxlib.library.messages.base.enums.QIDType;
import com.v2soft.styxlib.library.messages.base.structs.StyxQID;
import com.v2soft.styxlib.library.messages.base.structs.StyxStat;
import com.v2soft.styxlib.library.server.ClientState;
import com.v2soft.styxlib.library.types.ULong;

/**
 * In-Memory directory
 * @author vshcryabets@gmail.com
 *
 */
public class MemoryStyxDirectory
extends MemoryStyxFile {
    private Map<ClientState, StyxByteBufferWriteable> mBuffersMap;
    private List<IVirtualStyxFile> mFiles;

    public MemoryStyxDirectory(String name) {
        super(name);
        mQID.setType(QIDType.QTDIR);
        mFiles = new LinkedList<IVirtualStyxFile>();
        mBuffersMap = new HashMap<ClientState, StyxByteBufferWriteable>();
    }

    @Override
    public int getMode() {
        return (int) (FileMode.Directory.getMode() | 0x01FF);
    }

    @Override
    public IVirtualStyxFile walk(Iterator<String> pathElements, List<StyxQID> qids) {
        if ( pathElements.hasNext() ) {
            String filename = pathElements.next();
            for (IVirtualStyxFile file : mFiles) {
                if ( file.getName().equals(filename)) {
                    qids.add(file.getQID());
                    return file.walk(pathElements, qids);
                }
            }
            return null;
        }
        return super.walk(pathElements, qids);
    }

    @Override
    public boolean open(ClientState client, int mode) throws IOException {
        boolean result = ((mode&0x0F) == ModeType.OREAD);
        if ( result ) {
            // prepare binary structure of the directory
            int size = 0;
            final List<StyxStat> stats = new LinkedList<StyxStat>();
            for (IVirtualStyxFile file : mFiles) {
                final StyxStat stat = file.getStat();
                size += stat.getSize();
                stats.add(stat);
            }
            // allocate buffer
            StyxByteBufferWriteable buffer = new StyxByteBufferWriteable(size);
            for (StyxStat state : stats) {
                state.writeBinaryTo(buffer);
            }
            mBuffersMap.put(client, buffer);
        }
        return result;
    }

    @Override
    public long read(ClientState client, byte[] outbuffer, ULong offset, long count) throws StyxErrorMessageException {
        if ( !mBuffersMap.containsKey(client)) StyxErrorMessageException.doException("This file isn't open");
        final IoBuffer buffer = mBuffersMap.get(client).getBuffer();
        int boffset = buffer.limit();
        if ( offset.asLong() > boffset ) return 0;
        buffer.position((int) offset.asLong());
        int bleft = buffer.remaining();
        if ( count > bleft ) {
            count = bleft;
        }
        buffer.get(outbuffer, 0, (int) count);
        return count;
    }

    @Override
    public void close(ClientState client) {
        // remove buffer
        mBuffersMap.remove(client);
    }

    /**
     * Add child file
     * @param file
     */
    public void addFile(IVirtualStyxFile file) {
        // TODO check! may be this folder already contains file with same name 
        mFiles.add(file);
    }

    @Override
    public int write(ClientState client, byte[] data, ULong offset)
            throws StyxErrorMessageException {
        StyxErrorMessageException.doException("Can't write to directory");
        return 0;
    }

    @Override
    public void onConnectionClosed(ClientState state) {
        for (IVirtualStyxFile file : mFiles) {
            file.onConnectionClosed(state);
        }
        close(state);
    }

    /**
     * Delete child file
     * @param file file that should be removed from this directory
     */
    public boolean deleteFile(IVirtualStyxFile file) {
        return mFiles.remove(file);
    }
}

File: native/V2StyxLib/lib/vfs/MemoryStyxDirectory.cpp
/*
 * MemoryStyxDirectory.cpp
 *
 *  Created on: May 20, 2012
 *      Author: vshcryabets@gmail.com
 */

#include "vfs/MemoryStyxDirectory.h"
#include "types.h"
#include "messages/base/structs/StyxQID.h"
#include <vector>
#include "string.h"
#include "io/StyxDataWriter.h"

MemoryStyxDirectory::MemoryStyxDirectory(std::string name):
	MemoryStyxFile(name) {
	mQID->setType(QTDIR);
	mStat->setMode(getMode());
}

MemoryStyxDirectory::~MemoryStyxDirectory() {
}
/**
 * @return file access mode
 */
int MemoryStyxDirectory::getMode() {
	return Directory | 0x01FF;
}
/**
 * Open file
 * @param mode
 * @throws IOException
 */
bool MemoryStyxDirectory::open(ClientDetails *client, int mode){
	bool result = ((mode&0x0F) == OREAD);
	if ( result ) {
		// prepare binary structure of the directory
		size_t size = 0;
		std::vector<StyxStat*> stats;
		for ( FileList::iterator it = mFiles.begin();
				it != mFiles.end(); it++ ) {
			StyxStat *stat = (*it)->getStat();
			size += stat->getSize();
			stats.push_back(stat);
		}
		// allocate buffer
		uint8_t* buffer = new uint8_t(size);
		StyxDataWriter writer(buffer);
		for ( std::vector<StyxStat*>::iterator it = stats.begin();
				it != stats.end(); it++ ) {
			(*it)->writeBinaryTo(&writer);
		}
		mBuffersMap.insert(std::pair<ClientDetails*, uint8_t*>(client, buffer));
	}
	return result;
}
/**
 * Close file
 * @param mode
 */
void MemoryStyxDirectory::close(ClientDetails *client) {
	// remove buffer
	ClientsMap::iterator item = mBuffersMap.find(client);
	if ( item != mBuffersMap.end() ) {
		delete item->second;
		mBuffersMap.erase(item);
	}
}
/**
 * Read from file
 * @param offset offset from begining of the file
 * @param count number of bytes to read
 * @return number of bytes that was readed into the buffer
 */
size_t MemoryStyxDirectory::read(ClientDetails *client, uint8_t* buffer, uint64_t offset, size_t count) {
	ClientsMap::iterator it = mBuffersMap.find(client);
	if ( it == mBuffersMap.end() ) {
		return -1; // TODO there we should send Rerror with message "This file isn't open"
	}
	uint8_t* preparedData = it->second;
	size_t dataSize = preparedData->getCapacity();
	if ( offset > dataSize ) return 0;
	size_t remaining = dataSize - offset;
	if ( count > remaining ) {
		count = remaining;
	}
	memcpy(buffer, preparedData->getBuffer()+offset, count);
	return count;
}
IVirtualStyxFile* MemoryStyxDirectory::walk(std::vector<StyxString*> *pathElements, std::vector<StyxQID*> *qids) {
	if ( pathElements->size() > 0 ) {
		StyxString* filename = *(pathElements->begin());
		for ( FileList::iterator iterator = mFiles.begin(); iterator < mFiles.end(); iterator++ ) {
			StyxString itemFileName = (*iterator)->getName();
			if ( itemFileName.compare(*filename) == 0 ) {
				pathElements->erase(pathElements->begin());
				qids->push_back((*iterator)->getQID());
				return (*iterator)->walk(pathElements, qids);
			}
		}
		return NULL;
	}
	return MemoryStyxFile::walk(pathElements, qids);
}
void MemoryStyxDirectory::onConnectionClosed(ClientDetails *state) {
	for ( FileList::iterator it = mFiles.begin();
			it != mFiles.end(); it++ ) {
		(*it)->onConnectionClosed(state);
	}
	close(state);
}
void MemoryStyxDirectory::addFile(IVirtualStyxFile *file) {
    mFiles.push_back(file);
}
